src/Libraries/Microsoft.Extensions.Diagnostics.Extra/Logging/JustInTimeRedactor.cs

```diff
@@ -19,23 +21,162 @@ internal sealed class JustInTimeRedactor : IResettable
+    private bool TryRedactWithDiscriminator(object? value, Span<char> destination, out int charsWritten, ReadOnlySpan<char> format, IFormatProvider? formatProvider,
+        out string result, bool generateResult)
     {
-        Value = null;
+        Span<char> workBuffer = stackalloc char[MaxStackAllocation];
+        int charsInWorkBuffer = 0;
+        char[]? a = null;
+
+        // first step is to acquire the actual text to be redacted
+
+#if NET6_0_OR_GREATER
+        bool done = false;
+        if (value is ISpanFormattable sf)
+        {
+            done = sf.TryFormat(workBuffer, out charsInWorkBuffer, format, formatProvider);
+        }
+
+        if (done)
+        {
+            int maxLenToRedact = charsInWorkBuffer + DiscriminatortSeparator.Length + _discriminator.Length + MaxDayOfYearLength;
+            if (maxLenToRedact > workBuffer.Length)
+            {
+                a = ArrayPool<char>.Shared.Rent(maxLenToRedact);
+                workBuffer.Slice(0, charsInWorkBuffer).CopyTo(a);
+                workBuffer = a;
+            }
+        }
+        else
+#endif
+#pragma warning disable S1199
+        {
+#pragma warning restore S1199
+            ReadOnlySpan<char> inputAsSpan = default;
+            if (value is IFormattable f)
+            {
+                inputAsSpan = f.ToString(format.Length > 0 ? format.ToString() : string.Empty, formatProvider).AsSpan();
+            }
+            else if (value is char[] c)
+            {
+                // An attempt to call value.ToString() on a char[] will produce the string "System.Char[]" and redaction will be attempted on it,
+                // instead of the provided array.
+                inputAsSpan = c.AsSpan();
+            }
+            else
+            {
+                var str = value?.ToString() ?? string.Empty;
+                inputAsSpan = str.AsSpan();
+            }
+
+            int maxLenToRedact = inputAsSpan.Length + DiscriminatortSeparator.Length + _discriminator.Length + MaxDayOfYearLength;
+            if (maxLenToRedact > workBuffer.Length)
+            {
+                a = ArrayPool<char>.Shared.Rent(maxLenToRedact);
+                workBuffer = a;
+            }
+
+            inputAsSpan.CopyTo(workBuffer);
+            charsInWorkBuffer = inputAsSpan.Length;
+        }
+
+        // next step is to add the discriminator
+
+        DiscriminatortSeparator.AsSpan().CopyTo(workBuffer.Slice(charsInWorkBuffer));
+        charsInWorkBuffer += DiscriminatortSeparator.Length;
+
+        _discriminator.AsSpan().CopyTo(workBuffer.Slice(charsInWorkBuffer));
+        charsInWorkBuffer += _discriminator.Length;
+
+        // final step is to invoke the actual redactor
+
+        var finalToBeRedacted = workBuffer.Slice(0, charsInWorkBuffer);
+        result = string.Empty;
+
+        int redactedLen = _redactor!.GetRedactedLength(finalToBeRedacted);
+        if (redactedLen > destination.Length)
+        {
+            if (generateResult)
```

Comment:
Do we generate result only when redacted length is greater than destination?

Reply:
We only generate the results when this function is called in the ToString path.

I don't like how this code is composed together, but it's always a pain when dealing with spans. It constrains how things can fit together.

Reply:
My question was rather about the fact that you don't check for `generateResult` (and don't set the `result`) in case `redactedLen <= destination.Length`

Reply:
Oh. I see. Thanks!